# Импорт библиотек и создание рабочего окружения

Модулем в Python называется некоторая небольшая программа.

Каждая программа может импортировать модуль и получить доступ к его классам, функциям и объектам. Нужно заметить, что модуль может быть написан не только на Python, а например, на C или C++.

## Подключение модуля из стандартной библиотеки

Подключить модуль можно с помощью инструкции `import`. Модули обычно подключаются в самом начале кода.

In [1]:
import os
print(os.getcwd())

C:\Users\Иван\Documents\python_developer\lecture_2


После ключевого слова `import` указывается название модуля.

Одной инструкцией можно подключить несколько модулей, хотя этого не рекомендуется делать, так как это снижает читаемость кода.

> Когда вы импортируете модуль, интерпретатор Python ищет этот модуль в следующих местах:
- Директория, в которой находится файл, в котором вызывается команда импорта
- Если модуль не найден, Python ищет в каждой директории, определенной в переменной окружения PYTHONPATH.
- Если и там модуль не найден, Python проверяет путь заданный по умолчанию
>
> Путь поиска модулей сохранен в системном модуле `sys` в переменной `path`. Переменная `sys.path` содержит все три вышеописанных места поиска модулей.

In [2]:
import time, random
print(time.time())
print(random.random())

1583945372.1565273
0.19285441488159405


После импортирования модуля его название становится переменной, через которую можно получить доступ к атрибутам модуля. Например, можно обратиться к константе e, расположенной в модуле math:

In [3]:
import math
print(math.e)

2.718281828459045


In [4]:
math.e = 3
math.e

3

In [5]:
import math
math.e

3

### Использование псевдонимов

Если название модуля слишком длинное, или оно вам не нравится по каким-то другим причинам, то для него можно создать псевдоним, с помощью ключевого слова `as`.

Теперь доступ ко всем атрибутам модуля math осуществляется только с помощью переменной m, а переменной math в этой программе уже не будет.

In [6]:
import math as m
print(m.e)

3


### Инструкция from

Подключить определенные атрибуты модуля можно с помощью инструкции from. Она имеет несколько форматов:

```
from <Название модуля> import <Атрибут 1> [ as <Псевдоним 1> ], [<Атрибут 2> [ as <Псевдоним 2> ] ...]
from <Название модуля> import *
```

Первый формат позволяет подключить из модуля только указанные вами атрибуты. Для длинных имен также можно назначить псевдоним, указав его после ключевого слова `as`.

In [7]:
from math import e, ceil as c
print(e)
print(c(4.6))

3
5


Импортируемые атрибуты можно разместить на нескольких строках, если их много, для лучшей читаемости кода:

In [8]:
from math import (sin, cos,
                  tan, atan)

Второй формат инструкции from позволяет почти все переменные из модуля. Для примера импортируем все атрибуты из модуля sys.

Следует заметить, что не все атрибуты будут импортированы. Если в модуле определена переменная `__all__` (список атрибутов, которые могут быть подключены), то будут подключены только атрибуты из этого списка. Если переменная `__all__` не определена, то будут подключены все атрибуты, не начинающиеся с нижнего подчёркивания. Кроме того, необходимо учитывать, что импортирование всех атрибутов из модуля может нарушить пространство имен главной программы, так как переменные, имеющие одинаковые имена, будут перезаписаны.

In [9]:
from sys import *
print(version)
print(version_info)

3.6.5 |Anaconda, Inc.| (default, Mar 29 2018, 13:32:41) [MSC v.1900 64 bit (AMD64)]
sys.version_info(major=3, minor=6, micro=5, releaselevel='final', serial=0)


## Создание модуля

Можно использовать как библиотеку и самостоятельно созданный файл с функциями. Он подключается так же, как и стандартные библиотеки - через имя файла.

При импортировании модуля его код выполняется полностью. То есть, если программа что-то печатает, то при её импортировании это будет напечатано. Этого можно избежать, если проверять, запущен ли скрипт как программа, или импортирован. Это можно сделать с помощью переменной `__name__`, которая определена в любой программе, и равна "`__main__`", если скрипт запущен в качестве главной программы, и имя, если он импортирован.

Пример такого кода:

In [10]:
def hello():
    print('Hello, world!')

def fib(n):
    a = b = 1
    for i in range(n - 2):
        a, b = b, a + b
    return b

if __name__ == "__main__":
    hello()
    for i in range(10):
        print(fib(i))

Hello, world!
1
1
1
2
3
5
8
13
21
34


In [1]:
import module

all run


In [13]:
module.__name__

'module'

## Система управления пакетами pip

Для установки пакетов Python будет использоваться pip. Это система управления пакетами, которая используется для установки пакетов из Python Package Index (PyPI). Скорее всего, если у вас уже установлен Python, то установлен и pip.

Проверка версии pip:

In [2]:
!pip --version

pip 18.0 from C:\Users\Иван\AppData\Roaming\Python\Python36\site-packages\pip (python 3.6)



> **Примечание:**
>
> *В зависимости от того, как установлен и настроен Python в системе, может потребоваться использовать pip3 вместо pip.*

Установка пакета:

`pip install tabulate`

Обновление пакета:

`pip install --upgrade tabulate`

Установка пакета конкретной версии:

`pip install tabulate==2.6.0`

Удаление пакета:

`pip uninstall tabulate`

По умолчанию pip устанавливает пакеты в системную директорию.

Опция `--user` позволяет установить пакет в домашнюю директорию пользователя от имени пользователя.

In [ ]:
!pip install --user mypackage

`pip show` выведет информацию о пакете.

In [ ]:
!pip show astor

`pip list` выведет список всех установленных в данном окружении пакетов

In [1]:
!pip list

Package                            Version            
---------------------------------- -------------------
-                                  0.13.1             
-cikit-image                       0.13.1             
absl-py                            0.8.1              
alabaster                          0.7.10             
anaconda-client                    1.6.14             
anaconda-navigator                 1.8.7              
anaconda-project                   0.8.2              
asn1crypto                         0.24.0             
astor                              0.7.1              
astroid                            1.6.3              
astropy                            3.0.2              
attrs                              18.1.0             
Babel                              2.5.3              
backcall                           0.1.0              
backports.shutil-get-terminal-size 1.0.0              
beautifulsoup4                     4.6.0              
bitarray  

You are using pip version 18.0, however version 20.0.2 is available.
You should consider upgrading via the 'python -m pip install --upgrade pip' command.


## Виртуальное окружение

Виртуальное окружение - это изолированное окружение среды, которое позволяет нам использовать определенные версии приложений.

Все пакеты, работающие с виртуальным окружением решают одну проблему - они позволяют проектам, которые имеют различные (и часто конфликтующие) зависимости, сосуществовать на одной системе.

Некоторые пакеты для работы с виртуальными окружениями:

- **Virtualenv**

`virtualenv` - стандартный пакет для работы с виртуальным окружением. Используется вместе с virtualenvwrapper для более удобной работы.

Установка: `pip install virtualenv`


- **Conda**

Conda - пакетный менеджер, который также позволяет нам создавать виртуальные окружения.

Начиная с Python 3.3 и 3.4, рекомендуемый способ создания виртуального пространства – это использование инструмента командной строки **pyvenv**, который также включен в инсталляцию Python 3 по умолчанию. Однако, в версии 3.6 и выше, вам нужен **python3 -m venv**.

Новая виртуальная среда в **venv** создается при помощи команды:

```
python3 -m venv env
```

Эта команда создает каталог под названием "env", структура каталога которого схожа со следующей:

```
├── bin
│   ├── activate
│   ├── activate.csh
│   ├── activate.fish
│   ├── easy_install
│   ├── easy_install-3.5
│   ├── pip
│   ├── pip3
│   ├── pip3.5
│   ├── python -> python3.5
│   ├── python3 -> python3.5
│   └── python3.5 -> /Library/Frameworks/Python.framework/Versions/3.5/bin/python3.5
├── include
├── lib
│   └── python3.5
│       └── site-packages
└── pyvenv.cfg
```

- **bin** – файлы, которые взаимодействуют с виртуальной средой;
- **include** – С-заголовки, компилирующие пакеты Python;
- **lib** – копия версии Python вместе с папкой "site-packages", в которой установлена каждая зависимость.

Чтобы использовать эти пакеты (или ресурсы) среды в изоляции, вам нужно «активировать» их, запустив скрипт **activate** в каталоге "bin":

```
source env/bin/activate
```

Приглашение командной строки теперь носит префикс вашей среды (в нашем случае – env). Это индикатор того, что env в данный момент активен, что в свою очередь говорит о том, что выполнимые файлы Python используют пакеты и настройки только этой среды.

```
(env) $
```

Находясь в одном из окружений, можно ставить пакеты через pip, как обычно:

```
pip install flask
```

Список зависимостей проекта принято сохранять в файл с именем requirements.txt:

```
pip3 freeze > requirements.txt
```

Этот подход позволяет одной командой установить все зависимости, необходимые проекту:

```
pip3 install -r requirements.txt
```

Команда **deactivate** возвращает назад в контекст «system».